### PR TITLE
enforce projectID from middleware

### PR DESCRIPTION
**Pull request by shivasishdas**
Closes https://github.com/razorpay/metro/issues/193

Have filed a bug as well on implementation: https://github.com/grpc-ecosystem/grpc-gateway/issues/2222

### CODE REVIEW

File: internal/interceptors/auth.go [hunk 1]
```diff
@@ -24,13 +24,17 @@ func UnaryServerAuthInterceptor(authFunc grpc_auth.AuthFunc) grpc.UnaryServerInt
...
 
-func getUserPassword(ctx context.Context) (user string, password []byte, err error) {
+func getUserPasswordProjectID(ctx context.Context) (user string, password []byte, projectID string, err error) {
 	md, ok := metadata.FromIncomingContext(ctx)
 	if !ok {
 		err = status.Error(codes.Unauthenticated, "could not parse incoming context")
 		return
 	}
 
+	if val := md.Get("uri"); val != nil && len(val) > 0 {
```
Comment (sunnyagg) [inline internal/interceptors/auth.go:34]: add comments, is val a slice ?

File: internal/interceptors/auth.go [hunk 2]
```diff
@@ -68,9 +72,25 @@ func secureCompare(expected, actual string) bool {
 	return false
 }
 
+// extractProjectIDFromURI : This function extracts the projectID from the HTTP request URI.
+// Example1: extractProjectIDFromURI("/v1/projects/project1/topics/t123") = project1
+// Example2: extractProjectIDFromURI("/v1/projects/project7/subscriptions/s123") = project7
+// Example3: extractProjectIDFromURI("/v1/admin/topic/t987") = ""
```
Comment (sunnyagg) [inline internal/interceptors/auth.go:78]: will this middleware be called for admin routes ? example 3 doesn't seem to be valid. this middleware should be only for client auth. admin auth can be a different middleware

File: internal/interceptors/auth.go [hunk 3]
```diff
@@ -68,9 +72,25 @@ func secureCompare(expected, actual string) bool {
...
+func extractProjectIDFromURI(uri string) string {
+	if uri == "" {
+		return ""
+	}
+
+	parts := strings.Split(uri, "/")
```
Comment (sunnyagg) [inline internal/interceptors/auth.go:84]: should we use some regex here
v1/projects/{id}/{anythingelse}

File: internal/interceptors/auth.go (hunk 2 above)
Comment (shivasishdas) [reply internal/interceptors/auth.go:78]: Middlewares are different. Both of them are using a common function to extract `getUserPasswordProjectID()` from the `authorization` header. This is needed for both the auth types.

File: internal/interceptors/auth.go (hunk 1 above)
Comment (shivasishdas) [reply internal/interceptors/auth.go:36]: added

**Review by sunnyagg — CHANGES_REQUESTED** [review]
Add tests, coverage is going down

**Comment by shivasishdas** [comment]
> Add tests, coverage is going down

Adding tests for AppAuth and AdminAuth

**Review by sunnyagg — APPROVED** [review]

**Pull request merged**